src/mlpack/core/optimizers/cmaes/cmaes_impl.hpp

```diff
@@ -0,0 +1,781 @@
+     
+    }
+  }
+
+
+  /**
+   * Adds the mutation sigma*B*(D*z).
+   * @param x Search space vector.
+   * @param eps Mutation factor.
+   */
+  template<typename funcType>
+  void CMAES<funcType>::addMutation(double* x, double eps)
+  {
+    for (int i = 0; i < N; ++i)
+      tempRandom[i] = rgD[i]*rand.gauss();
+    for (int i = 0; i < N; ++i)
+    {
+      double sum = 0.0;
+      for (int j = 0; j < N; ++j)
+        sum += B(i,j)*tempRandom[j];
+      x[i] = xmean[i] + eps*sigma*sum;
+    }
+  }
+
+  /**
+   * Initializes the CMA-ES algorithm.
+   * @param parameters The CMA-ES parameters in the parameters.h file
+   * @return Array of size lambda that can be used to assign fitness values and
+   *         pass them to updateDistribution()
+   */
+  template<typename funcType>
+  void CMAES<funcType>::init(arma::vec& func)
+  {
+
+    stopMessage = "";
+
+    double trace = arma::accu(arma::pow(rgInitialStds, 2));
+    sigma = std::sqrt(trace/N);
+
+    chiN = std::sqrt((double) N) * (1 - 1/(4*N) + 1/(21*N*N));
+    eigensysIsUptodate = true;
+    doCheckEigen = false;
+    genOfEigensysUpdate = 0;
+
+    double dtest;
+    for (dtest = double(1); dtest && dtest < double(1.1)*dtest; dtest *= double(2))
+      if (dtest == dtest + double(1))
+        break;
+    dMaxSignifKond = dtest / double(1000);
+
+    gen = 0;
+    countevals = 0;
+    state = INITIALIZED;
+    dLastMinEWgroesserNull = double(1);
+
+    pc.set_size(N);
+    ps.set_size(N);
+    tempRandom.set_size(N+1);
+    BDz.set_size(N);
+    xmean.set_size(N+2);
+    xmean[0] = N;
+    ++xmean;
+    xold.set_size(N+2);
+    xold[0] = N;
+    ++xold;
+    xBestEver.set_size(N+3);
+    xBestEver[0] = N;
+    ++xBestEver;
+    xBestEver[N] = std::numeric_limits<double>::max();
+    rgD.set_size(N);
+    C.set_size(N,N);
+    B.set_size(N,N);
+    publicFitness.set_size(lambda);
+    functionValues = new double[lambda+1];
+    functionValues[0] = lambda;
+    ++functionValues;
+    const int historySize = 10 + (int) ceil(3.*10.*N/lambda);
+    funcValueHistory = new double[historySize + 1];
+    funcValueHistory[0] = (double) historySize;
+    funcValueHistory++;
+    index = arma::linspace<arma::vec>(0, lambda-1, lambda);
+    population.zeros(lambda, N+2);
+    for (int i = 0; i < lambda; i++)
+    {
+      functionValues[i] = std::numeric_limits<double>::max();
+    }
+    for (int i = 0; i < historySize; i++)
+    {
+      funcValueHistory[i] = std::numeric_limits<double>::max();
+    }
+    
+    C.zeros();
+    B.zeros();
+
+    for (int i = 0; i < N; ++i)
+    {
+      B.diag().ones();
+
+      C(i,i) = rgD[i] = rgInitialStds[i]*std::sqrt(N/trace);
+      C(i,i) *= C(i,i);
+      pc[i] = ps[i] = double(0);
+    }
+    minEW = rgD.min();
+    minEW = minEW*minEW;
+    maxEW = rgD.max();
+    maxEW = maxEW*maxEW;
+
+    maxdiagC = arma::max(C.diag());
+    mindiagC = arma::min(C.diag());
+
+    for (int i = 0; i < N; ++i)
+      xmean[i] = xold[i] = xstart[i];
```

Comment:
In line 381 you set ``xmean[0] = N`` and here you reset the value, not sure doing the first step is necessary, since we can just use ``xmean.n_elem`` to get the size.

Also, we can just do:

```c++
xold = xstart;
xmean = xold;
```